File: Komodo.Core/Compilation/AST/Identifier.cs
using Komodo.Core.Utilities;
using Komodo.Core.Compilation.TypeSystem;

namespace Komodo.Core.Compilation.AST;

public record Identifier : INode
{
    public string Name { get; }
    public Symbol Symbol { get; }
    public TextLocation Location { get; }

    private Identifier(string name, TextLocation location, Symbol symbol)
    {
        Name = name;
        Location = location;
        Symbol = symbol;
    }

    public NodeType NodeType => NodeType.Identifier;
    public INode[] Children => new INode[] { };

    public record Expression(string Name, TextLocation Location, Symbol Symbol) : Identifier(Name, Location, Symbol), IExpression
    {
        public TSType TSType => Symbol switch
        {
            Symbol.Variable var => var.TSType,
            Symbol.Function func => func.TSType,
            var symbol => throw new NotImplementedException($"Identifier expressions can not contain symbols of type {symbol.GetType()}") 
        };
    }

    public record Typename(string Name, TextLocation Location, Symbol.Typename TSTypename) : Identifier(Name, Location, TSTypename) { }
}

File: src/compiler/ast/expression/identifier.rs
use serde::Serialize;

use crate::compiler::{ast::Node, typesystem::ts_type::TSType, utilities::location::Location};

#[derive(Serialize)]
pub struct Identifier<'source> {
    value: String,
    ts_type: TSType,
    location: Location<'source>,
}

impl<'source> Identifier<'source> {
    pub fn new(value: String, ts_type: TSType, location: Location<'source>) -> Self {
        Self {
            value,
            ts_type,
            location,
        }
    }

    pub fn value(&self) -> &str {
        &self.value
    }
}

impl<'source> Node<'source> for Identifier<'source> {
    fn ts_type(&self) -> &TSType {
        &self.ts_type
    }

    fn location(&self) -> Location<'source> {
        self.location.clone()
    }
}
